src/usn.rs

```diff
@@ -0,0 +1,331 @@
+#[cfg(feature = "multithreading")]
+use rayon;
+use std::io;
+use regex::bytes;
+use std::cmp::max;
+use std::fs::File;
+use std::io::SeekFrom;
+#[cfg(feature = "multithreading")]
+use rayon::prelude::*;
+use byteorder::{ByteOrder, LittleEndian};
+use crate::ReadSeek;
+use crate::record::UsnEntry;
+
+// This is the size of data chunks
+const SIZE_CHUNK: usize = 17408;
+// This is the size of the search within a chunk
+// It is smaller than the chunk size to garentee that the last record found is complete
+// It has been noticed that generally usn records are paged in 4096 byte pages. I have not
+// observed usn records overlaping the 4096 offset and are zero padded to the 4096 mark.
+const SIZE_SEARCH: usize = 16384;
+
+
+pub struct UsnParserSettings{
+    thread_count: usize
+}
+
+impl Default for UsnParserSettings {
+    fn default() -> Self {
+        UsnParserSettings {
+            thread_count: 0
+        }
+    }
+}
+
+impl UsnParserSettings {
+    pub fn new() -> UsnParserSettings {
+        UsnParserSettings::default()
+    }
+
+    /// Sets the number of worker threads.
+    /// `0` will let rayon decide.
+    ///
+    #[cfg(feature = "multithreading")]
+    pub fn thread_count(mut self, thread_count: usize) -> Self {
+        self.thread_count = if thread_count == 0 {
+            rayon::current_num_threads()
+        } else {
+            thread_count
+        };
+        self
+    }
+
+    /// Does nothing and emits a warning when complied without multithreading.
+    #[cfg(not(feature = "multithreading"))]
+    pub fn thread_count(mut self, _thread_count: usize) -> Self {
+        warn!("Setting num_threads has no effect when compiling without multithreading support.");
+        self.thread_count = 1;
+        self
+    }
+}
+
+
+pub struct UsnParser<T: ReadSeek> {
+    _inner_handle: T,
+    _size: u64,
+    settings: UsnParserSettings
+}
+
+impl UsnParser<File> {
+    pub fn from_path(filename: &str) -> Result<Self, io::Error> {
+        let file_handle = File::open(filename)?;
+
+        Self::from_read_seek(
+            file_handle
+        )
+    }
+}
+
+impl <T: ReadSeek> UsnParser <T> {
+    pub fn from_read_seek(mut inner_handle: T) -> Result<Self, io::Error> {
+        // We need to get the end offset to determine the size
+        let end_offset = inner_handle.seek(SeekFrom::End(0))?;
+
+        // Seek back to the beginning
+        inner_handle.seek(SeekFrom::Start(0))?;
+
+        Ok( Self {
+            _inner_handle: inner_handle,
```

Comment:
Contrary to python, the rust convention is to use visibility modifier (e.g `pub` when having public attributes, otherwise private), no need for the _ prefix.